### PR TITLE
fix VS2015 - second attempt

### LINKED ISSUE #66: [vs2015] Error loading in VS 2015
```
    <type>Error</type>
    <source>VisualStudio</source>
    <description>CreateInstance failed for package [VSConanPackage]Source: &apos;mscorlib&apos; Description: Could not load file or assembly &apos;Microsoft.VisualStudio.Shell.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a&apos; or one of its dependencies. The system cannot find the file specified.&#x000D;&#x000A;System.IO.FileNotFoundException: Could not load file or assembly &apos;Microsoft.VisualStudio.Shell.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a&apos; or one of its dependencies. The system cannot find the file specified.&#x000D;&#x000A;File name: &apos;Microsoft.VisualStudio.Shell.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a&apos;&#x000D;&#x000A;   at System.Reflection.RuntimeAssembly.GetType(RuntimeAssembly assembly, String name, Boolean throwOnError, Boolean ignoreCase, ObjectHandleOnStack type)&#x000D;&#x000A;   at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase)&#x000D;&#x000A;   at System.Activator.CreateInstanceFromInternal(String assemblyFile, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo)&#x000D;&#x000A;   at System.AppDomain.CreateInstanceFrom(String assemblyFile, String typeName)&#x000D;&#x000A;&#x000D;&#x000A;WRN: Assembly binding logging is turned OFF.&#x000D;&#x000A;To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.&#x000D;&#x000A;Note: There is some performance penalty associated with assembly bind failure logging.&#x000D;&#x000A;To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].&#x000D;&#x000A;</description>
    <guid>{33315C89-72DD-43BB-863C-561C1AA5ED54}</guid>
    <hr>80004005 - E_FAIL</hr>
    <errorinfo></errorinfo>
```

Maybe related: https://github.com/samizzo/nshader/issues/12
Also: https://github.com/StefanKert/BuildVision/issues/34

Note.- It has to be tested **in a machine with only 2015 installed** (other installations of VS in the same machine hide this error).

### LINKED ISSUE #98: Think alternatives for VS2015
Current efforts trying to make the extension compatible with VS2015 have arrived at a dead-end. There is an incompatibility regarding `VCProjectEngine.dll` for different versions of the library that is shipped with different versions of Visual Studio and the casting from `Project.object` to `VCProject` fails for one version or the other.

Here there are some useful links that are interesting to read: https://github.com/conan-io/conan-vs-extension/pull/75#issuecomment-505465228

I'm opening this issue to think about alternatives and decide the roadmap for it.

Thanks

### PR DESCRIPTION
**Pull request by SSE4**
closes #98
closes #66
closes #75

extension doesn't work in Visual Studio 2015 - it's because interface `Microsoft.VisualStudio.VCProject` has changed between versions `14.0` and `15.0` and it's no longer compatible.
unfortunately, it's impossible to reference both versions of the same interface within single assembly.  it's also impossible to solve it somehow in runtime, as at compile time expression like `project as VCProject` compiles differently depends on `VCProject` interface version (or more precisely, its guid).
the solution is pretty straightforward: 
1. wrap all code needed from `Microsoft.VisualStudio.VCProject` and `Microsoft.VisualStudio.VCProjectEngine` into small wrapper classes. for now, just 3 classes are needed: `VCConfiguration`, `VCProject` and `VCPropertySheet`. (other classes might be easily added on demand using the same approach)
2. define corresponding interfaces within `Core` project, so classes are never referenced directly, and we can use late binding for them, so they are loosely coupled.
3. create two wrapper assemblies for two Visual Studio versions: `VCProjectWrapper14` and `VCProjectWrapper15`. both assemblies use the same sources (3 wrapper classes mentioned above), but reference to different versions of `Microsoft.VisualStudio.VCProject`
4. in main assembly, load required wrapper assembly based on the Visual Studio version in use. from the loaded assembly, create an instance of `VCProjectWrapper` (from the `EnvDTE.Project.object`) and use it going forward. 
5. ensure we're not referencing `Microsoft.VisualStudio.VCProject` (and `VCProjectEngine`) from the main assembly.
6. downgrade all Visual-Studio related assembly references to the 14.0, so they could be loaded by Visual Studio 2015
7. use `Microsoft.VisualStudio.VCProject` (and `VCProjectEngine`) version 14.0 from NuGet, as we're building on VS2019 image which doesn't have older assemblies installed.

### CODE REVIEW

**Comment by SSE4** [comment]
seems like build fails because assemblies like `Microsoft.VisualStudio.VCProject, Version=14.0.0.0` aren't found on appveyor...

**Comment by jgsogo** [comment]
Look at this wonderful image,  Version: 14.0.25431 😍😍😍

image

Please, rebase on top of `dev` and now the Appveyor wars will start again ^^

File: Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Conan.VisualStudio.VCProjectWrapper")]
+[assembly: AssemblyDescription("")]
```
Comment (jgsogo) [inline Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyDescription("Conan extension for Visual Studio")]
```

File: Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
```
Comment (jgsogo) [inline Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyCompany("JFrog LTD")]
```

File: Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+[assembly: AssemblyProduct("Conan.VisualStudio.VCProjectWrapper")]
```
Comment (jgsogo) [inline Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyProduct("Conan.VisualStudio")]
```

File: Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+[assembly: AssemblyCopyright("Copyright ©  2019")]
```
Comment (jgsogo) [inline Conan.VisualStudio.VCProjectWrapper/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyCopyright("Copyright (c) 2019 JFrog LTD")]
```

File: Conan.VisualStudio/Services/VcProjectService.cs [hunk 5]
```diff
@@ -36,21 +34,48 @@ public bool IsConanProject(Project project)
             return IsCppProject(project) && null != ConanPathHelper.GetNearestConanfilePath(AsVCProject(project).ProjectDirectory);
         }
 
-        public VCProject AsVCProject(Project project)
+        private static IVCProject CreateVCProjectWrapper(Project project)
+        {
+            int version = ConanCompilerVersion();
```
Comment (jgsogo) [inline Conan.VisualStudio/Services/VcProjectService.cs]: I feel like `ConanCompilerVersion` is not the proper name to retrieve the version of VS (More on this on the review comment)

Comment (SSE4) [reply Conan.VisualStudio/Services/VcProjectService.cs]: I gonna rename it

Comment (SSE4) [reply Conan.VisualStudio/Services/VcProjectService.cs]: renamed to VisualStudioVersion (property)

File: vsix/check_installed.ps1 [hunk 6]
```diff
@@ -16,7 +16,7 @@ $ntimes = 5
 For ($i=1; $i -le $ntimes; $i++) {  # Run 10 times
     $output = . "$devenv" /ConanVisualStudioVersion /?
 
-    $pattern = "^${env:APPVEYOR_BUILD_VERSION}\s+Microsoft Visual Studio"  # Version + output from /? command
+    $pattern = "^${env:APPVEYOR_BUILD_VERSION}\s+"  # Version + output from /? command
```
Comment (jgsogo) [inline vsix/check_installed.ps1]: Is this change still needed?

Comment (SSE4) [reply vsix/check_installed.ps1]: yes, because of extra output from devenv.exe (warnings)

Comment (SSE4) [reply vsix/check_installed.ps1]: anyway, it's not a big deal, as we're going to verify by actual integration test rather than by just version check

**Pull request merged**